rasterio/_io.pyx

```diff
@@ -1801,3 +1801,85 @@ def virtual_file_to_buffer(filename):
     log.debug("Buffer length: %d bytes", n)
     cdef np.uint8_t[:] buff_view = <np.uint8_t[:n]>buff
     return buff_view
+
+
+
+cdef class WarpedVRTReaderBase(DatasetReaderBase):
+
+    def __init__(self, name, dst_crs=None, resampling=Resampling.nearest,
+                 tolerance=0.15, ):
+        super(WarpedVRTReaderBase, self).__init__(self)
+        self.name = name
```

Comment:
I would argue that `name` is not entirely truthful here, but I'm not sure what we should use as a replacement.  The dataset only exists in memory, so (according to the `GDALAutoCreateWarpedVRT()` [docs](http://www.gdal.org/gdalwarper_8h.html#ab5a8723d68786e7554f1ad4c0a6fa8d3)) we could give it a new name with `GDALSetDescription()`.  Maybe something like `<WarpedVRT:tests/data/RGB.byte.tif>`, which is similar to what `__repr__()` returns for a class it cannot precisely describe?

Just using the input name suggests the data you read is the same as though you opened `name` directly, but thats not true.  For example:

```python
import rasterio as rio
from rasterio.io import VirtualWarpedFile


with VirtualWarpedFile('tests/data/RGB.byte.tif', dst_crs='EPSG:4326').open() as warped:
    with rio.open(warped.name) as src, rio.open('tests/data/RGB.byte.tif') as truth:
        print("Pixel match: {}".format(warped.read() == src.read() == truth.read()))
        print("Name match: {}".format(warped.name == src.name == truth.name))
        print("Shape match: {}".format(warped.shape == src.shape == truth.shape))
```